File: engine_py/tests/test_ai_job_assignment.py
import pytest


def test_assign_jobs_basic(make_world):
    world = make_world()
    agent = world.spawn_entity()
    world.set_component(
        agent, "Agent", {"entity_id": agent, "skills": {"TestJob": 1.0}}
    )

    job_entity = world.spawn_entity()
    world.assign_job(
        job_entity,
        "TestJob",
        state="pending",
        category="test",
        assigned_to=None,
    )

    jobs_before = world.list_jobs()
    assert any(
        j["id"] == job_entity
        and (not j.get("assigned_to") or j.get("assigned_to") == 0)
        for j in jobs_before
    )

    world.ai_assign_jobs(agent, [])

    jobs_after = world.list_jobs()
    assigned_jobs = [j for j in jobs_after if j.get("assigned_to") == agent]
    assert any(
        j["id"] == job_entity for j in assigned_jobs
    ), "Job should be assigned to agent"


def test_assign_jobs_multiple_agents(make_world):
    world = make_world()

    agent1 = world.spawn_entity()
    world.set_component(
        agent1, "Agent", {"entity_id": agent1, "skills": {"TestJob": 1.0}}
    )

    agent2 = world.spawn_entity()
    world.set_component(
        agent2, "Agent", {"entity_id": agent2, "skills": {"TestJob": 1.0}}
    )

    job1 = world.spawn_entity()
    world.assign_job(job1, "TestJob", state="pending", category="test")

    job2 = world.spawn_entity()
    world.assign_job(job2, "TestJob", state="pending", category="test")

    world.ai_assign_jobs(agent1, [])
    world.ai_assign_jobs(agent2, [])

    jobs = world.list_jobs()
    assigned_jobs = [
        j for j in jobs if j["state"] in ("pending", "in_progress")
    ]
    assert len(assigned_jobs) == 2

    assigned_counts = {}
    for j in assigned_jobs:
        assigned_to = j.get("assigned_to")
        assigned_counts[assigned_to] = assigned_counts.get(assigned_to, 0) + 1

    count1 = assigned_counts.get(agent1, 0)
    count2 = assigned_counts.get(agent2, 0)
    diff = abs(count1 - count2)
    assert diff <= 1


def test_assign_jobs_no_jobs(make_world):
    world = make_world()

    agent = world.spawn_entity()
    world.set_component(
        agent, "Agent", {"entity_id": agent, "skills": {"TestJob": 1.0}}
    )

    # No jobs created here
    world.ai_assign_jobs(agent, [])

    jobs = world.list_jobs()
    assert len(jobs) == 0

File: engine/scripts/lua/tests/test_ai_job_assignment.lua
local assert = require("assert")
local ai_helpers = require("helpers.ai_job_helpers")
local job_helpers = require("helpers.job_helpers") -- if needed

local function test_assign_jobs_basic()
	init_job_event_logger()
	local agent = ai_helpers.spawn_agent_with_skills({ TestJob = 1.0 })

	local job_entity = spawn_entity()
	assign_job(job_entity, "TestJob", { state = "pending", category = "test", assigned_to = nil })

	local jobs_before = list_jobs()
	local found_unassigned = false
	for _, j in ipairs(jobs_before) do
		if j.id == job_entity and (not j.assigned_to or j.assigned_to == 0) then
			found_unassigned = true
		end
	end
	assert.is_true(found_unassigned, "Job should initially be unassigned")

	ai_assign_jobs(agent, {})

	local jobs_after = list_jobs()
	local found_assigned = false
	for _, j in ipairs(jobs_after) do
		if j.id == job_entity and j.assigned_to == agent then
			found_assigned = true
		end
	end
	assert.is_true(found_assigned, "Job should be assigned to the agent by AI logic")
end

local function test_assign_jobs_multiple_agents()
	local agent1 = ai_helpers.spawn_agent_with_skills({ TestJob = 1.0 })
	local agent2 = ai_helpers.spawn_agent_with_skills({ TestJob = 1.0 })

	local job1 = spawn_entity()
	assign_job(job1, "TestJob", { state = "pending", category = "test" })

	local job2 = spawn_entity()
	assign_job(job2, "TestJob", { state = "pending", category = "test" })

	ai_assign_jobs(agent1, {})
	ai_assign_jobs(agent2, {})

	local jobs = list_jobs()
	local assigned_jobs = 0
	local assigned_counts = {}
	for _, j in ipairs(jobs) do
		if j.state == "pending" or j.state == "in_progress" then
			assigned_jobs = assigned_jobs + 1
			assigned_counts[j.assigned_to] = (assigned_counts[j.assigned_to] or 0) + 1
		end
	end

	assert.equals(2, assigned_jobs, "Both jobs should be assigned")

	local count1 = assigned_counts[agent1] or 0
	local count2 = assigned_counts[agent2] or 0
	local diff = math.abs(count1 - count2)
	assert.is_true(diff <= 1, "Jobs should be fairly assigned among agents")
end

local function test_assign_jobs_no_jobs()
	local agent = ai_helpers.spawn_agent_with_skills({ TestJob = 1.0 })

	ai_assign_jobs(agent, {})

	local jobs = list_jobs()
	assert.equals(0, #jobs, "No jobs should exist in world")
end

return {
	test_assign_jobs_basic = test_assign_jobs_basic,
	test_assign_jobs_multiple_agents = test_assign_jobs_multiple_agents,
	test_assign_jobs_no_jobs = test_assign_jobs_no_jobs,
}
